lib/AppInfo/Application.php

```diff
@@ -25,26 +25,35 @@
 use OCA\External\Settings\Personal;
 use OCA\External\SitesManager;
 use OCP\AppFramework\App;
+use OCP\AppFramework\Bootstrap\IBootContext;
+use OCP\AppFramework\Bootstrap\IBootstrap;
+use OCP\AppFramework\Bootstrap\IRegistrationContext;
+use OCP\INavigationManager;
 use OCP\IServerContainer;
+use OCP\IURLGenerator;
+use OCP\Settings\IManager;
+use OCP\Util;
 use Symfony\Component\EventDispatcher\GenericEvent;
 
-class Application extends App {
+class Application extends App implements IBootstrap {
 
-	public function __construct() {
-		parent::__construct('external');
+	public const APP_ID = 'external';
 
-		$this->getContainer()->registerCapability(Capabilities::class);
+	public function __construct() {
+		parent::__construct(self::APP_ID);
 	}
 
-	public function register() {
-		$server = $this->getContainer()->getServer();
+	public function register(IRegistrationContext $context): void {
+		$context->registerCapability(Capabilities::class);
+	}
 
+	public function boot(IBootContext $context): void {
 		/** @var SitesManager $sitesManager */
-		$sitesManager = $this->getContainer()->query(SitesManager::class);
+		$sitesManager = $context->getAppContainer()->get(SitesManager::class);
 		$sites = $sitesManager->getSitesToDisplay();
 
-		$this->registerNavigationEntries($server, $sites);
-		$this->registerPersonalPage($server, $sites);
+		$this->registerNavigationEntries($context->getServerContainer(), $sites);
```

Comment:
may I suggest the use of `\OCP\AppFramework\Bootstrap\IBootContext::injectFn` to get rid of the explicit service locator?

Reply:
Ah, nice thing.

Yeah I guess this hole code should be refactored into a listener for BeforeTemplateRenderedEvent any way

Reply:
Okay, doesn't work for the settings module as the event is too late. But I can not (yet) use injectFn() there, due to the missing alias: https://github.com/nextcloud/server/pull/22597 unless I use non public API